File: php/anonymous.php
<?php

declare(strict_types=1);

namespace Acme\Synthetic;

function makeMultiplier(int $factor): \Closure
{
    return function (int $value) use ($factor): int {
        return $value * $factor;
    };
}

function makeAdder(int $offset): callable
{
    return fn(int $value): int => $value + $offset;
}

function applyAll(array $values, callable $transform): array
{
    $out = [];
    foreach ($values as $value) {
        $out[] = $transform($value);
    }
    return $out;
}

function compose(callable ...$callables): \Closure
{
    return function (mixed $input) use ($callables): mixed {
        $result = $input;
        foreach ($callables as $callable) {
            $result = $callable($result);
        }
        return $result;
    };
}

function buildAnonymousCalculator(int $base): object
{
    return new class($base) {
        public function __construct(private readonly int $base) {}

        public function add(int $other): int
        {
            return $this->base + $other;
        }

        public function multiply(int $other): int
        {
            return $this->base * $other;
        }

        public function summarize(array $values): array
        {
            $sum = array_reduce($values, fn(int $a, int $b): int => $a + $b, 0);
            $product = array_reduce(
                $values,
                function (int $acc, int $value): int {
                    return $acc * $value;
                },
                1,
            );
            return ['sum' => $sum, 'product' => $product];
        }
    };
}

function partitionBy(array $values, callable $predicate): array
{
    $matched = [];
    $rest = [];
    foreach ($values as $value) {
        if ($predicate($value)) {
            $matched[] = $value;
        } else {
            $rest[] = $value;
        }
    }
    return [$matched, $rest];
}

File: csharp/anonymous.cs
using System;
using System.Collections.Generic;

namespace Acme.Synthetic
{
    public static class AnonymousSamples
    {
        public static Func<int, int> MakeMultiplier(int factor)
        {
            return value => value * factor;
        }

        public static Func<int, int> MakeAdder(int offset)
        {
            Func<int, int> adder = delegate(int value) { return value + offset; };
            return adder;
        }

        public static List<int> ApplyAll(List<int> values, Func<int, int> transform)
        {
            var output = new List<int>();
            foreach (var value in values)
            {
                output.Add(transform(value));
            }
            return output;
        }

        public static Func<int, int> Compose(params Func<int, int>[] callables)
        {
            return input =>
            {
                var result = input;
                foreach (var callable in callables)
                {
                    result = callable(result);
                }
                return result;
            };
        }

        public static int SumWithLocal(List<int> values)
        {
            int Add(int a, int b) => a + b;

            int Total(IEnumerable<int> xs)
            {
                var acc = 0;
                foreach (var x in xs)
                {
                    acc = Add(acc, x);
                }
                return acc;
            }

            return Total(values);
        }

        public static (List<int> matched, List<int> rest) PartitionBy(
            List<int> values,
            Func<int, bool> predicate)
        {
            var matched = new List<int>();
            var rest = new List<int>();
            foreach (var value in values)
            {
                if (predicate(value))
                {
                    matched.Add(value);
                }
                else
                {
                    rest.Add(value);
                }
            }
            return (matched, rest);
        }
    }
}
